File: maid/ai/schemas.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import List, Optional

from mako.template import Template


@dataclass
class CompletionError:
    error: str


class Role(str, Enum):
    user = "user"
    assistant = "assistant"
    system = "system"


class FinishReason(str, Enum):
    stop = "stop"
    length = "length"
    content_filter = "content_filter"
    null = "null"


@dataclass
class CompletionTokens:
    prompt: int
    completion: int
    total: int


@dataclass
class CompletionMessage:
    role: Role
    content: str


@dataclass
class CompletionChoice:
    message: CompletionMessage
    finish_reason: Optional[FinishReason]
    index: int


@dataclass
class CompletionResult:
    usage: CompletionTokens
    choices: List[CompletionChoice]


@dataclass
class Prompt:
    role: Role
    content: str

    def join(self, with_: Prompt) -> Prompt:
        return Prompt(
            role=self.role, content=f"{self.content}.\n\n{with_.content}"
        )

    def render(self, *pos, **kws) -> Prompt:
        return Prompt(
            role=self.role,
            content=Template(self.content).render(*pos, **kws),
        )

    @classmethod
    def assistant(cls, content: str) -> Prompt:
        return Prompt(role=Role.assistant, content=content)

    @classmethod
    def system(cls, content: str) -> Prompt:
        return Prompt(role=Role.system, content=content)

    @classmethod
    def user(cls, content: str) -> Prompt:
        return Prompt(role=Role.user, content=content)

File: packages/maid-openai/src/schemas.rs
use serde::{
    Deserialize,
    Serialize,
};

#[derive(
    Debug, Serialize, Clone, Copy, PartialEq, Eq, Deserialize,
)]
#[serde(rename_all = "snake_case")]
pub enum FinishReason {
    Stop,
    Length,
    ContentFilter,
}

#[derive(
    Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize,
)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    User,
    Assistant,
    System,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Prompt {
    pub role: Role,
    pub content: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TokensUsage {
    #[serde(rename = "prompt_tokens")]
    pub prompt: u32,

    #[serde(rename = "completion_tokens")]
    pub completion: u32,

    #[serde(rename = "total_tokens")]
    pub total: u32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AnswerMessage {
    pub role: Role,
    pub content: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Answer {
    Error {
        error: ErrorAnswer,
    },
    Success {
        #[serde(flatten)]
        result: SuccessfulAnswer,
    },
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ErrorAnswer {
    pub message: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SuccessfulAnswer {
    pub usage: TokensUsage,
    pub choices: Vec<AnswerChoice>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AnswerChoice {
    pub message: AnswerMessage,
    pub finish_reason: Option<FinishReason>,
}

impl Prompt {
    pub fn new(role: Role, content: impl Into<String>) -> Self {
        Self {
            role,
            content: content.into(),
        }
    }

    pub fn assistant(content: impl Into<String>) -> Self {
        Self {
            role: Role::Assistant,
            content: content.into(),
        }
    }

    pub fn system(content: impl Into<String>) -> Self {
        Self {
            role: Role::System,
            content: content.into(),
        }
    }

    pub fn user(content: impl Into<String>) -> Self {
        Self {
            role: Role::User,
            content: content.into(),
        }
    }
}
